crates/aranya-daemon/src/sync/manager.rs

```diff
@@ -0,0 +1,1219 @@
+use core::error;
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use aranya_daemon_api::SyncPeerConfig;
+use aranya_runtime::{Address, GraphId, SyncHelloType, SyncType, MAX_SYNC_MESSAGE_SIZE};
+use aranya_util::ready;
+use dashmap::DashMap;
+use futures_util::StreamExt as _;
+use tokio::{
+    sync::mpsc,
+    task::{self, AbortHandle, JoinError, JoinSet},
+};
+use tokio_util::time::{delay_queue::Key, DelayQueue};
+use tracing::error;
+
+use super::SyncPeer;
+
+#[derive(Debug, thiserror::Error)]
+enum SyncError {
+    #[error("This peer hasn't been registered yet")]
+    PeerNotRegistered,
+    #[error("There's already an active sync for this peer")]
+    AlreadyActive,
+    #[error("Unable to send manager a message")]
+    ManagerShutdown,
+    #[error("The buffer for this peer is already being used")]
+    BufferInUse,
+    #[error(transparent)]
+    Postcard(#[from] postcard::Error),
+}
+type Result<T> = core::result::Result<T, SyncError>;
+
+enum ManagerMessage {
+    /// Add a peer to the sync schedule.
+    AddPeer {
+        peer: SyncPeer,
+        config: SyncPeerConfig,
+    },
+    /// Remove a peer from the sync schedule.
+    RemovePeer { peer: SyncPeer },
+    /// Request to sync with a peer now (bypasses `DelayQueue`).
+    SyncNow { peer: SyncPeer },
+    /// A [`HelloService`]-specific message.
+    Hello(HelloMessage),
+}
+
+enum HelloMessage {
+    /// Subscribe to hello notifications from this peer.
+    Subscribe { peer: SyncPeer, config: HelloConfig },
+    /// Unsubscribe from hello notifications from this peer.
+    Unsubscribe { peer: SyncPeer },
+
+    /// A peer wants to subscribe to hello notifications.
+    PeerSubscribed { peer: SyncPeer, config: HelloConfig },
+    /// A peer wants to unsubscribe from hello notifiactions.
+    PeerUnsubscribed { peer: SyncPeer },
+
+    /// Send a hello notification to this peer.
+    SendNotification { peer: SyncPeer, head: Address },
+    /// Received a hello notification from a peer.
+    HelloReceived { peer: SyncPeer, head: Address },
+    /// Our graph head changed, so notify all subscribed peers.
+    GraphHeadChanged { graph_id: GraphId, head: Address },
+}
+
+// TODO(nikki): not happy having to keep 3 enums synced (ActiveSync, TaskResult, ProtocolMessage).
+enum ActiveSync {
+    /// Sync message that requires a response.
+    Message {
+        started_at: Instant,
+        handle: AbortHandle,
+    },
+    /// Single request that expects an ack.
+    Request {
+        started_at: Instant,
+        handle: AbortHandle,
+    },
+}
```

Comment:
This may be simpler than an enum since the fields are the same. The only difference is whether they receive an ack or response:
```suggestion
struct ActiveSync {
    /// When the sync request started.
    started_at: Instant,
    /// Sync abort handle.
    handle: AbortHandle,
    /// Requires a response.
    response: bool,
}
```

Reply:
It's mostly about if we care about the additional information of what kind of sync it is, I can definitely flatten it.